create ext_hdd variable and refactor

gui1117 opened the pull request:
ext_hdd is true when on armhf or armv7 and an external drive exist

refactor: system always mount external hard drive if ext_hdd

Commit:
create ext_hdd variable and refactor

ext_hdd is true when on armhf or armv7 and an external drive exist

refactor: system always mount external hard drive if ext_hdd

thiolliere commented:
not finished yet have to add to ansible local fact

gui1117 commented:
for now this branch implements ext_hdd=if on armv7 or armhf and if sda1 is block device.

If I put it to installed_software.fact then ext_hdd will be determine on the first run.

It will fail if the external hard drive no longer exist

I think it would be better if there is an optionnal extra var to ansible that is ext_hdd=/dev/sda1 (if we want the external drive to be /dev/sda1)

Commit:
write ext_hdd only on first start

fheslouin commented:
> for now this branch implements ext_hdd=if on armv7 or armhf and if sda1 is block device.

This sounds good to me ! 

> If I put it to installed_software.fact then ext_hdd will be determine on the first run.

Yes, we think : ok if a box has a device attached, it will be always the same case. Otherwise we have to think about how to migrate the data from HDD to Sdcard and I think this is a no go.

> It will fail if the external hard drive no longer exist

Yes it will, same as if the sd card or HDD is damaged

> I think it would be better if there is an optionnal extra var to ansible that is ext_hdd=/dev/sda1 (if we want the external drive to be /dev/sda1)

That is something we used to do by the past, but I think this was not so reliable, It can be quite dangerous to handle such variable from the command line. I would rather choose something that we can automate, but it is also nice to give choice to the end-user. 
This could be discussed with @letompouce  and @swbsf

gui1117 commented:
When I used this branch, the two last tasks of set_custom_fact are not even played ...

I don't understand why. Emmanuel has the same error https://framagit.org/ideascube/pibox-installer/issues/15

Commit:
remove scalwey-arm empty file

Commit:
tags always

gui1117 commented:
it must be because it was untagged

fheslouin commented on roles/mook/tasks/back_end.yml:
```diff
@@ -24,23 +24,15 @@
     
-- name: Create static dir
-  file: path={{ mook_path }}/{{ mook_name }}/static/tmp state=directory owner={{username}} group={{username}} mode=0775
-  when: (ansible_architecture == 'x86_64' or ansible_architecture == 'i386')
```
This task is used for x86 arch, if you delete it no static dir will be created

gui1117 replied:
my fault I didn't read the code correctly

fheslouin commented on roles/mook/tasks/back_end.yml:
```diff
@@ -24,23 +24,15 @@
-  tags:
-    - custom
-    
+  tags: ['custom']
```
In your case, the static dir will be created always on /media/hdd, but what happen if no external HDD as on x86 arch ?

fheslouin commented on roles/set_custom_fact/tasks/main.yml:
```diff
@@ -45,22 +45,6 @@
 - block:
 
   - name: Generate a random number for the idc_uniq_device_name
-    set_fact: idc_uniq_device_name="{{ ideascube_project_name }}"
```
in this task we are not adding random number at the end, because we are at master run. We want to generate a standard ansiblecube device

gui1117 replied:
ok, I drop this commit as it can and should be done in #122

fheslouin replied:
ok great

fheslouin commented on roles/set_custom_fact/tasks/main.yml:
```diff
@@ -45,22 +45,6 @@
-- block:
-
-  - name: Generate a random number for the idc_uniq_device_name
     set_fact: idc_uniq_device_name="{{ ideascube_project_name }}_{{ 1000 | random }}"
```
In this task `custom` and `rename` run, if we give an `ideascube_project_name` we add a random number as we can have more than one device with this name.

gui1117 replied:
ok, I drop this commit as it can and should be done in #122

fheslouin commented on roles/set_custom_fact/tasks/main.yml:
```diff
@@ -145,3 +129,14 @@
+
+- name: Use sda1 as external hard drive if sda1 exists and on arm architecture
+  set_fact: ext_hdd={{ (ansible_architecture == "armhf" or ansible_architecture == "armv7l") and (sda1.stat.isblk | default (omit)) == True }}
+  when: firststart.stat.exists == false or firststart is undefined
```
If you use this condition, this task will be executed only when `/etc/firststart.csv` does not exist, meaning when the device has no device name defined. However `ext_hdd` is going to be used through the whole playbook, this will lead to an error. 

Either you store the `ext_hdd` variable in custom fact in order to be accessible later on in the playbook or you remove the condition.

gui1117 replied:
doesn't set_fact store the variable in custom fact ? how can I do it otherwise ?

fheslouin replied:
About set_fact 
> These variables will be available to subsequent plays during an ansible-playbook run, but will not be saved across executions even if you use a fact cache.

So it can work if you remove the `when:` condition.

Otherwise, you have to store the `ext_hdd` in a fact file as done here https://github.com/ideascube/ansiblecube/blob/oneUpdateFile/roles/set_custom_fact/tasks/main.yml#L18-L22

We have two options : 

1. We "always" test if an HDD is attached to the device and tasks will be executed depending the result of `ext_hdd`, this way it could be a little bit more "dynamic" and we could image to switch from sd storage to hdd storage. Data still have to be migrated somehow.
2. We test only one time at custom run if an HDD is attached to the device. We store the result as a fact in a fact file `/etc/ansible/facts.d/installed_software.fact` and we use it as it is. The device will have always the same state. 

I think option 1 will not work anyway because if an HDD exist it has to be registered in `/etc/fstab`, if once the device start without HDD it will lead to an issue as the attached HDD define in fstab won't be found. There is no use then of testing at each run this variable. Storing in fact file is the best option from my point of view @letompouce @swbsf what do you think ?

Commit:
set ext_hdd variable for each run

there were two options:
    We "always" test if an HDD is attached to the device and tasks will be executed depending the result of ext_hdd, this way it could be a little bit more "dynamic" and we could image to switch from sd storage to hdd storage. Data still have to be migrated somehow.
    We test only one time at custom run if an HDD is attached to the device. We store the result as a fact in a fact file /etc/ansible/facts.d/installed_software.fact and we use it as it is. The device will have always the same state.

this commit implement option 1

fheslouin committed:
Merge branch 'oneUpdateFile' into create_ext_hdd_variable
